src/test/java/com/google/sps/TripServletTest.java

```diff
@@ -0,0 +1,131 @@
+import com.google.maps.DirectionsApiRequest;
+import com.google.maps.GeoApiContext;
+import com.google.maps.model.DirectionsResult;
+import com.google.maps.model.DirectionsRoute;
+import com.google.maps.model.DirectionsLeg;
+import com.google.maps.model.Duration;
+import com.google.maps.model.TravelMode;
+import com.google.sps.data.Config;
+import com.google.sps.servlets.TripServlet;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentMatchers;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import static org.mockito.Mockito.*;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(DirectionsApi.class)
+public final class TripServletTest {
+
+  // Test directionsRequest generation with mocks.
+  @Test
+  public void generateDirectionsRequestTest() {
+    PowerMockito.mockStatic(DirectionsApi.class);
+    DirectionsApiRequest mockRequest = mock(DirectionsApiRequest.class);
+    when(DirectionsApi.newRequest(any())).thenReturn(mockRequest);
+    when(mockRequest.origin(anyString())).thenReturn(mockRequest);
+    when(mockRequest.destination(anyString())).thenReturn(mockRequest);
+    when(mockRequest.waypoints(ArgumentMatchers.<String>any())).thenReturn(mockRequest);
+    when(mockRequest.optimizeWaypoints(anyBoolean())).thenReturn(mockRequest);
+    when(mockRequest.mode(any())).thenReturn(mockRequest);
+
+    GeoApiContext mockGeoApiContext = mock(GeoApiContext.class);
+
+    String[] waypoints = new String[]{ "MoPOP, 5th Avenue North, Seattle, WA, USA",
+                                    "Space Needle, Broad Street, Seattle, WA, USA",
+                                    "Alki Beach, Seattle, WA, USA"};
+    
+    String origin = "The Westin Bellevue, Bellevue Way Northeast, Bellevue, WA, USA";
+
+    DirectionsApiRequest request = TripServlet.generateDirectionsRequest(origin, waypoints, mockGeoApiContext);
+
+    // Verify that proper methods and parameters are called to generate directionsRequest.
+    verify(mockRequest).origin(origin);
+    verify(mockRequest).destination(origin);
+    verify(mockRequest).waypoints(waypoints);
+    verify(mockRequest).optimizeWaypoints(true);
+    verify(mockRequest).mode(TravelMode.DRIVING);
+  }
+
+  // Test waypoint order parsing from a DirectionsResult object.
+  @Test
+  public void getOrderedWaypointsTest() {
+    // Manually create directions result object.
+    DirectionsResult dirResult = new DirectionsResult();
+    dirResult.routes = new DirectionsRoute[1];
+    dirResult.routes[0] = new DirectionsRoute();
+    dirResult.routes[0].waypointOrder = new int[]{ 1, 2, 0 };
```

Comment:
I don't really know whats going on here. whats the difference between `DirectionsRoute[..]` and ``DirectionsRoute()`. Looks complicated. You can either type it out or hop on a call at some point?